### PR TITLE
.NET Standard support!

### PR DESCRIPTION
**Pull request by ViktorHofer**
This PR contains tons of project changes to adapt to netstandard and netcore. The .NET Core team is currently looking into popular .NET projects and their potential for migration to .NET Standard. This contribution is part of that initiative and should help boosting the adoption of .NET Standard.

I tried to contact Brian Lagunas (via blog) and Brian Noyes (via email) to align with them regarding the contribution but did not got a response yet. Therefore I continued my work and now created a PR.

1. Migration from PCL to netstandard1.0:
- Prism
- Prism.Forms
- Prism.Autofac.Forms
- Prism.DryIoc.Forms
- Prism.Ninject.Forms
- Prism.Unity.Forms

2. Migration of net45 test projects to netcore1.1:
- Prism.Tests
- Prism.Forms.Tests
- Prism.Autofac.Forms.Tests
- Prism.DryIoc.Forms.Tests
- Prism.Unity.Forms.Tests

3. Removal of packages.config, Properties folder and nuspec file for all netstandard and netcore libraries.

4. Fixed project references in some solution files.

5. Enabled automatic nupkg creation on build for all netstandard and netcore libraries.

6. AppVeyor DevOps scenarios with publishing to myget

### CODE REVIEW

**Comment by bartlannoeye** [comment]
Hey @ViktorHofer, the plan was to wait until .NETStandard 2.0 (see https://github.com/PrismLibrary/Prism/issues/683#issuecomment-263600123). I'll let @brianlagunas decide what to do, as you obviously did put quite some effort into it.

File: README.md [hunk 1]
```diff
@@ -89,11 +89,11 @@ Installation is straightforward if you've installed Xamarin Add-ins before, just
 ![Xamarin Studio Addin Manager](docs/images/prism-xamarin-studio-addin-manager.jpg)
 
 # Samples
-We have both a development sandbox (frequently changing) and stable samples for using Prism with WPF, UWP and Xamarin Forms. An overview of the samples can be found [here](Sandbox/README.md).
+We have both a development sandbox (frequently changing) and stable samples for using Prism with WPF, UWP and Xamarin Forms. An overview of the samples can be found [here](samples/README.md).
```
Comment (bartlannoeye) [inline README.md]: Url incorrect? Must be sandbox/readme.md?

**Comment by ViktorHofer** [comment]
Oh didn't know that and didn't find any info about that in the docs. Thanks for the link! Yeah it was some effort, especially because Xamarin.Forms had a build bug and first that issues had to be fixed: https://bugzilla.xamarin.com/show_bug.cgi?id=53060

The netstandard libs in my PR are all targeting version 1.0 therefore they support the most environments possible with netstandards (WP8, WP81, Win81, UWP; netcore, net45, ...).

As I'm also planning to put some effort into Unity (if that's welcomed), it would be best to align before I start.

**Comment by bartlannoeye** [comment]
Don't think Unity is still actively maintained (and accepting PRs).

**Comment by bartlannoeye** [comment]
Shouldn't all files from the .github folder stay where they were? Or am I missing something that allows them to be moved to the /docs tree?
https://github.com/blog/2111-issue-and-pull-request-templates

**Comment by ViktorHofer** [comment]
Yeah your are totally right. I did that at late night and I'm not convinced at all that those changes would improve anything so I will change the files back into the .github directory and also rename the src directory back to Source. Than appveyor should also work with that PR and I won't produce any breaking changes

File: src/Prism/Mvvm/ErrorsContainer.cs [hunk 2]
```diff
@@ -26,12 +24,7 @@ public class ErrorsContainer<T>
         /// event.</param>
         public ErrorsContainer(Action<string> raiseErrorsChanged)
         {
-            if (raiseErrorsChanged == null)
-            {
-                throw new ArgumentNullException(nameof(raiseErrorsChanged));
-            }
-
-            this.raiseErrorsChanged = raiseErrorsChanged;
+            this.raiseErrorsChanged = raiseErrorsChanged ?? throw new ArgumentNullException(nameof(raiseErrorsChanged));
```
Comment (bartlannoeye) [inline src/Prism/Mvvm/ErrorsContainer.cs]: private fields should start with a _, so this won't be necessary if the fields follow the guidance.

**Comment by ViktorHofer** [comment]
Changes are reverted back now. AppVeyor is still failing as its configuration needs to be adjusted (new csproj format).

File: README.md (hunk 1 above)
Comment (ViktorHofer) [inline README.md]: Fixed

**Comment by bartlannoeye** [comment]
That's something @brianlagunas will have to do, I'll leave the rest to him.

**Comment by brianlagunas** [comment]
None of the unit tests run because VS does not support testing .NET Standard.

**Comment by ViktorHofer** [comment]
Hey Brian,

what do you mean with 

> None of the unit tests run because VS does not support testing .NET Standard.

I just tested to run the unit tests and 840/880 passed and the other 40 didn't start correctly. The issue that those tests do not run does not seem to be connected to this PR.

**Comment by brianlagunas** [comment]
The VS Runner does not support .NET Standard Projects.  You converted the test projects to .NET Standard, and now those test will not run.  We have more than 880 tests :).

You can test this by opening the PrismLibrary_Core.sln and try running the Prism.Tests.

**Comment by ViktorHofer** [comment]
I did not converted the test projects to netstandard but to netcore1.1. I just pushed an update so that the netcore projects now produce executable and the vs test runner now detects the unit tests.

Running the tests in Prism.Tests is working for me from within Visual Studio or via dotnet test command.

**Comment by brianlagunas** [comment]
I am unable to run the Tests and they are not showing in Test Explorer.  I get no test artifacts in the code editor.

Screenshot

**Comment by ViktorHofer** [comment]
It does work me

netcore-test

Visual Studio 2017 is currently bit buggy with xunit test execution and discovery. net core tools --> "dotnet test" works fine.

**Comment by ViktorHofer** [comment]
cc: @bradwilson, @onovotny

**Comment by ViktorHofer** [comment]
Others are also seeing this error: http://stackoverflow.com/questions/42747868/vs-2017-not-discovering-tests-if-net-core

**Comment by ViktorHofer** [comment]
@brianlagunas I'm currently finishing up my changes and hope to be ready by the end of tomorrow. I'm setting up the appveyor yml configuration so that it supports WPF, UWP and netcore/netstandard. As I want to align with you how to proceed to accept the PR please send me an email as I wasn't able to find your contact: viktor.hofer@microsoft.com.

**Comment by ViktorHofer** [comment]
@brianlagunas @bartlannoeye @dansiegel PR should be ready for review. Xamarin Forms bug resolved and unit tests are visible in VS2017 (newest update!). 35 tests are failing but I think this is because the new Xamarin.Forms version changed something. Please review!

**Comment by brianlagunas** [comment]
I upgraded my VS to the latest and the tests still don't work.  I also noticed that you completely changed the folder structure of the projects, renamed solutions and folders, and essentially changed a lot of solution structure.  That's like going into my garage and moving all my tools without telling me.  What was your motivating factor for this?

**Comment by brianlagunas** [comment]
The reason the tests are failing is because it appears you removed the FormsApplication class that was used for testing and moved it into the Prism.DI.Forms.Tests project: https://github.com/PrismLibrary/Prism/blob/master/Source/Xamarin/Prism.Forms/FormsApplication.cs

This breaks all XF tests.

There are a ton of changes in this PR, some are structural that I don't like.  This is very high risk PR.

**Comment by ViktorHofer** [comment]
```
The reason the tests are failing is because it appears you removed the FormsApplication class that was used for testing and moved it into the Prism.DI.Forms.Tests project: https://github.com/PrismLibrary/Prism/blob/master/Source/Xamarin/Prism.Forms/FormsApplication.cs
This breaks all XF tests.
```

Ooh that wasn't intentional. The rebasing wasn't that easy because of 30+ new commits so I might have made mistakes.

**Comment by brianlagunas** [comment]
> See it more as showing you a picture how your garage could look.

You're obviously not a car guy :)

**Comment by bradwilson** [comment]
>> Edit: I have tests showing in the Test Explorer in VS2017, but the test indicator is not showing in the code files.
> Yeah that's a known bug in combination with netcore and should be fixed soon.

That should be fixed in 2.3 beta 1, which we just shipped today. https://xunit.github.io/release-notes/2017-04-06.html

**Comment by ViktorHofer** [comment]
@bradwilson: Thanks Brad! Updated to 2.3 beta 1 and CodeLenses are indeed showing up again 👍 

@brianlagunas: Found the issue. By migrating to netcoreapp/netstandard projects I forgot to add the TEST condition. Is in again and now all unit tests are passing and you have the green/red indicators (CodeLens) working again.

**Comment by wocar** [comment]
Will there ever be support for netstandard 🥇  Cheers :)

**Comment by dansiegel** [comment]
@wocar Yes... it's being worked on

**Comment by ericwj** [comment]
Is the plan to also clean up and standardize the DI infrastructure around IServiceProvider and Microsoft.Extensions.DependencyInjection.Abstractions? This has the potential in fact to remove the strict need for any of the DI provider specific packages. App developers could add any available one on top of a reference to Prism.Windows for example by referencing e.g. 'Autofac.Extensions.DependencyInjection'. No need for overrides of any PrismApplication methods apart from the constructor - at least I think so.

**Comment by dansiegel** [comment]
@ericwj Can you open a new issue fully describing the API that you're thinking of. I'm not sure if it's something that Prism will adopt, but I do believe it would be worth having a conversation about. 

The conversion to NetStandard is mostly focused on delivering better support for Cross Platform apps (Xamarin Forms) which will also allow the updating of Autofac and Ninject to their current releases.

**Comment by weitzhandler** [comment]
I'm working on XF project and now need a nuget package that's based on standard 1.3. The only reason I keep Xamarin as PCL is because I'm tied to Prism. Are we gonna see standard support before 2.0?

**Comment by brianlagunas** [comment]
Adding .NET Standard Support isn't a simple tasks and requires a ton of effort.  We are looking into this now, and will release a preview when we think we are ready for testing.

**Comment by weitzhandler** [comment]
@brianlagunas really appreciate your kind response, and thanks for all the great effort put into this can't-live-without project!

**Comment by wocar** [comment]
How ever you can reference Prism in your netstandard library without
trouble and use it in your Xamarin Forms app in android and iOS

**Comment by dansiegel** [comment]
@wocar you can try out the preview templates I have available on [MyGet](https://myget.org/gallery/dansiegel-templates). They work with dotnet new and use the NetStandard core. In order to use PCL packages in your NetStandard library your project needs to be using the new csproj format, and you need to add an override to the `PackageTargetFallback` which those templates do for you. If you have more questions I'm happy to help you further on the Slack Channel

**Comment by weitzhandler** [comment]
This issue is starting to become really serious to me. I have a Xamarin PCL and I can't add some libraries that already support .NET Standard because I'm tied to Prism.

@dansiegel I didn't really understand, is that myget package only templates? Do you have a feed that has the netstandard-supported Prism.* libraries?

**Comment by dansiegel** [comment]
@weitzhandler that nuget contains project and item templates that can be created using dotnet new. The project template has a NetStandard1.4 core project and works just fine with Prism 6.3.0. You do not new Prism to convert over to NetStandard to use it with a NetStandard project. If you need more I'd be happy to explain further on [Slack](https://prismslack.herokuapp.com/).

**Comment by brianlagunas** [comment]
This is an issue in general and not just Prism.  Prism actually has an issue upgrading because some of the containers that don't support .NET Standard.  You may have to use an earlier version of your packages that support the PCL profile.  Even Xamarin.Forms itself does not support .NET Standard, and won't until the end of the year.  This is something we are working on, but it's not a simple change to make.  Microsoft really created a big cluster-fu*k with this.

**Comment by weitzhandler** [comment]
@brianlagunas found a PCL version of the lib I was looking for[*](https://www.nuget.org/packages/IdentityModel/1.13.1), thanks for this direction.

**Pull request closed without merging**